src/python/pants/core/goals/typecheck.py

```diff
@@ -41,18 +43,69 @@ def prep_output(s: bytes) -> str:
             exit_code=process_result.exit_code,
             stdout=prep_output(process_result.stdout),
             stderr=prep_output(process_result.stderr),
-            typechecker_name=typechecker_name,
+            partition_description=partition_description,
         )
 
 
-class TypecheckResults(Collection[TypecheckResult]):
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class TypecheckResults(EngineAware):
     """Zero or more TypecheckResult objects for a single type checker.
 
     Typically, type checkers will return one result. If they no-oped, they will return zero results.
     However, some type checkers may need to partition their input and thus may need to return
     multiple results.
     """
 
+    results: Tuple[TypecheckResult, ...]
+    typechecker_name: str
+
+    def __init__(self, results: Iterable[TypecheckResult], *, typechecker_name: str) -> None:
+        self.results = tuple(results)
+        self.typechecker_name = typechecker_name
+
+    @property
+    def skipped(self) -> bool:
+        return bool(self.results) is False
+
+    @memoized_property
+    def exit_code(self) -> int:
+        return next((result.exit_code for result in self.results if result.exit_code != 0), 0)
+
+    def level(self) -> Optional[LogLevel]:
+        if self.skipped:
+            return LogLevel.DEBUG
+        return LogLevel.WARN if self.exit_code != 0 else LogLevel.INFO
```

Comment:
The reason this method can return None (iirc) is that if you don't return a level, the workunit completes with the level it started with... ie, whatever the `@rule` is marked with. So I don't think it makes sense to always return a non-None value here, as that will override your `@rule(level=..)`... and maybe result in something logging that it is `[INFO] Starting` and then `[DEBUG] Completed`... which would be very odd.

So it feels like this should be:
```
def level(self) -> Optional[LogLevel]:
    return LogLevel.WARN if self.exit_code != 0 else None
```

Reply:
> The reason this method can return None (iirc) is that if you don't return a level, the workunit completes with the level it started with... ie, whatever the `@rule` is marked with. So I don't think it makes sense to always return a non-None value here, as that will override your `@rule(level=..)`... and maybe result in something logging that it is `[INFO] Starting` and then `[DEBUG] Completed`... which would be very odd.
> 
> So it feels like this should be:
> 
> ```
> def level(self) -> Optional[LogLevel]:
>     return LogLevel.WARN if self.exit_code != 0 else None
> ```

Yes, the idea behind the methods on `EngineAware` is that if they return None, they won't change the engine's behavior from what it was going to do anyway. The default implementations of all the methods on the `EngineAware` class itself do nothing and return `None`.

Reply:
In fact, I just realized that https://github.com/pantsbuild/pants/blob/master/src/python/pants/core/goals/test.py#L116 does the same thing, and I think this is why I'm seeing Completed messages at INFO level when the corresponding Started messages are at DEBUG level, and I'm going to fix this in a separate PR now.

Reply:
This is intentional for a better UX. 

See https://github.com/pantsbuild/pants/issues/10536. Until we solve https://github.com/pantsbuild/pants/issues/7907, those starting log messages would be really noisy. Because they can't have dynamic info, all they would say is "Run Pytest". If you run 100 tests, you would have "Run Pytest" 100 times.

Once we solve #7907, then this could be changed so that we can say "Started - Run Pytest for foo".

Reply:
@Eric-Arellano makes sense, disregard my previous comments.